BranchMetrics/android-branch-deep-linking-attribution
# Sdk 1090 fix timeout

## Reference
https://branch.atlassian.net/browse/SDK-1090

## Description
Currently the SDK does not actually respect the value passed in the API, `setNetworkTimeout`, because of the `HttpsURLConnection.setConnectTimeout(...)` implementation that applies the passed in timeout to all IP addresses that an endpoint resolves to. We host `v1/open` endpoint on 10+ machines, so in a typical case with our default timeout of 5.5. seconds, we will not timeout for like a full minute.

Applied solution:
Artificially cap the allocated time given to the request to complete, and cancel the request otherwise. This is the solution I'm planning to implement for right now.

## Testing Instructions
Run the tester app, try to set the network timeout to various values and observe timeouts happen at the expected time. Here is a command to run emulator with super high latency to repro timeouts:
`emulator -avd NameOfYouEmulator -netspeed 1:1 -netdelay 500`

## Risk Assessment [`HIGH` || `MEDIUM` || `LOW`]
MEDIUM


## Reviewer Checklist (To be checked off by the reviewer only)

- Correctness & Style
